### PR TITLE
Hive: Arrange common part of the code for Iceberg View.

### PR DESCRIPTION
**Pull request by nk1506**
Hive considers iceberg-view just another type of Hive-Table. So Iceberg-View on Hive should have same strategy for commit as Iceberg-Table on Hive. As part of this refactoring all the common code for table which can be used by view too has been moved to common places. Now `HiveOperationsBase` has all the common code for committing the iceberg-table and iceberg-view.  

This change also define a new abstract class `BaseMetastoreOperations`   which can contains common code from `BaseMetastoreTableOperations` and `BaseViewOperations` .

Motivation behind this change to have less diff for implementing Iceberg-view on Hive.

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/BaseMetastoreTableOperations.java [hunk 1]
```diff
@@ -309,65 +304,39 @@ protected enum CommitStatus {
    * @return Commit Status of Success, Failure or Unknown
    */
   protected CommitStatus checkCommitStatus(String newMetadataLocation, TableMetadata config) {
-    int maxAttempts =
```
Comment (nk1506) [inline core/src/main/java/org/apache/iceberg/BaseMetastoreTableOperations.java:312]: Since Hive-View also needs to check the commit status, so it has been moved to an abstract class `BaseMetastoreOperations`.

File: core/src/main/java/org/apache/iceberg/CatalogUtil.java [hunk 2]
```diff
@@ -136,6 +138,18 @@ public static void dropTableData(FileIO io, TableMetadata metadata) {
...
 
+  /**
+   * Drops view metadata files referenced by ViewMetadata.
+   *
+   * <p>This should be called by dropView implementations
+   *
+   * @param io a FileIO to use for deletes
+   * @param metadata the last valid ViewMetadata instance for a dropped view.
+   */
+  public static void dropViewMetadata(FileIO io, ViewMetadata metadata) {
```
Comment (nk1506) [inline core/src/main/java/org/apache/iceberg/CatalogUtil.java:149]: This change can go as part of Hive-View implementation.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HMSParametersSetter.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.hive;
+
+import java.util.Set;
+import org.apache.hadoop.hive.metastore.api.Table;
+
+/** An interface to set the required HMS parameters to the table. */
+interface HMSParametersSetter {
```
Comment (nk1506) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HMSParametersSetter.java:25]: Iceberg table on Hive and Iceberg-view on Hive needs to set few parameters like metadatalocation, schema, etc. Few of the metrics needs to be fetched from TableMetadata/ViewMetadata. Since we don't have a common abstract class for both. This consumer can help to set the respective parameters with their own implementation .

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 4]
```diff
@@ -222,24 +214,46 @@ public boolean dropTable(TableIdentifier identifier, boolean purge) {
...
+        .filter(
+            table ->
+                table.getParameters() != null
+                    && tableTypeProp.equalsIgnoreCase(
+                        table.getParameters().get(BaseMetastoreTableOperations.TABLE_TYPE_PROP)))
+        .map(table -> TableIdentifier.of(namespace, table.getTableName()))
+        .collect(Collectors.toList());
+  }
+
+  private void renameContent(
```
Comment (nk1506) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:235]: This can be used by view for `renameView` .

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 5]
```diff
@@ -250,29 +264,66 @@ public void renameTable(TableIdentifier from, TableIdentifier originalTo) {
...
+      String catalogName,
+      Table table,
+      TableIdentifier identifier) {
+    switch (contentType) {
+      case TABLE:
+        HiveOperationsBase.validateTableIsIceberg(
+            table, CatalogUtil.fullTableName(catalogName, identifier));
+        break;
+      case VIEW:
+        throw new UnsupportedOperationException("View is not supported.");
```
Comment (nk1506) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:307]: This can be updated with Hive-View .

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveOperationsBase.java [hunk 6]
```diff
@@ -181,4 +272,203 @@ default Table newHmsTable(String hmsTableOwner) {
 
     return newTable;
   }
+
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
+  default void commitWithHiveLock(
```
Comment (nk1506) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveOperationsBase.java:277]: This is a common code for committing `Iceberg-Table` and `Iceberg-View `on `Hive`.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 7]
```diff
@@ -166,168 +159,36 @@ protected void doRefresh() {
     refreshFromMetadataLocation(metadataLocation, metadataRefreshMaxRetries);
   }
 
-  @SuppressWarnings("checkstyle:CyclomaticComplexity")
   @Override
   protected void doCommit(TableMetadata base, TableMetadata metadata) {
     boolean newTable = base == null;
     String newMetadataLocation = writeNewMetadataIfRequired(newTable, metadata);
-    boolean hiveEngineEnabled = hiveEngineEnabled(metadata, conf);
```
Comment (nk1506) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:174]: All this common code has been moved to `HiveOperationsBase` .

File: core/src/main/java/org/apache/iceberg/BaseMetastoreTableOperations.java [hunk 8]
```diff
@@ -309,65 +304,39 @@ protected enum CommitStatus {
...
+
+  /**
+   * Checks the new metadata location presents or not after refreshing the table.
+   *
+   * @param newMetadataLocation newly written metadata location
+   * @return true if the new metadata location presents with current or previous metadata files.
+   */
+  protected boolean checkCurrentMetadataLocation(String newMetadataLocation) {
+    TableMetadata metadata = refresh();
+    Preconditions.checkNotNull(metadata, "Unexpected null table metadata");
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/BaseMetastoreTableOperations.java:324]: I would have expected this code to be what the previous code was:

```
TableMetadata metadata = refresh();
String currentMetadataFileLocation = metadata.metadataFileLocation();
boolean commitSuccess =
   currentMetadataFileLocation.equals(newMetadataLocation)
   || metadata.previousFiles().stream().anyMatch(log -> log.file().equals(newMetadataLocation));
```

can you elaborate why the code here now ends up being different?

File: core/src/main/java/org/apache/iceberg/CatalogUtil.java (hunk 2 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/CatalogUtil.java:149]: from what I understand, this PR only refactors a bunch of things to make it easier to add view support. That means we probably shouldn't be adding view-specific things in this PR as we just want to move stuff around and provide the exact same code behavior that previously existed, right?

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 9]
```diff
@@ -222,24 +214,46 @@ public boolean dropTable(TableIdentifier identifier, boolean purge) {
...
       throw new NoSuchNamespaceException(
-          "Cannot rename %s to %s. Namespace does not exist: %s", from, to, to.namespace());
+          "Cannot rename %s to %s. Namespace does not exist: %s",
+          fromIdentifier, to, to.namespace());
     }
 
     String toDatabase = to.namespace().level(0);
-    String fromDatabase = from.namespace().level(0);
-    String fromName = from.name();
+    String fromDatabase = fromIdentifier.namespace().level(0);
```
Comment (nastra) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:251]: why is the rename required here?

Comment (nastra) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:251]:
```suggestion
    String fromDatabase = from.namespace().level(0);
```

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 10]
```diff
@@ -250,29 +264,66 @@ public void renameTable(TableIdentifier from, TableIdentifier originalTo) {
...
+          throw new NoSuchViewException(
+              "Cannot rename %s to %s. View does not exist", fromIdentifier, to);
+      }
 
     } catch (InvalidOperationException e) {
       if (e.getMessage() != null
           && e.getMessage().contains(String.format("new table %s already exists", to))) {
-        throw new org.apache.iceberg.exceptions.AlreadyExistsException(
-            "Table already exists: %s", to);
+        throwErrorForExistedToContent(fromIdentifier, removeCatalogName(toIdentifier));
```
Comment (nastra) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:282]: I don't think we want to do this here, as this would do another call to retrieve the table. I would leave the code here as it was before

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveOperationsBase.java [hunk 11]
```diff
@@ -169,8 +260,8 @@ default Table newHmsTable(String hmsTableOwner) {
             null,
             Collections.emptyList(),
             Maps.newHashMap(),
-            null,
-            null,
+            properties.getOrDefault(HiveCatalog.VIEW_ORIGINAL_TEXT, null),
```
Comment (nastra) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveOperationsBase.java:263]: why is this adding view-specific things?

Comment (nk1506) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveOperationsBase.java:263]: I think this might not be specific to Hive-View.  This is another field in Hive-Table .  IMO, This field either we can ignore for both iceberg-table and view or make it optional with properties. 
Ref: https://github.com/apache/hive/blame/df45194268ffb10f9f7aae0ab4e3aec35a7b31d8/standalone-metastore/metastore-common/src/gen/thrift/gen-javabean/org/apache/hadoop/hive/metastore/api/Table.java#L23

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java (hunk 10 above)
Comment (nk1506) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:282]: Motivation behind this changes is  
`ViewCatalogTests` expects error message like `Cannot rename %s to %s. %s already exists`. It should explicitly tell the upstream **TO** tableType. 

Also we are making only one call to retrieve TO content. 
Nevertheless, What do you propose to match the error messages for  `ViewCatalogTests` ?

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveOperationsBase.java (hunk 11 above)
Comment (nastra) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveOperationsBase.java:263]: I think this PR should be a pure refactoring and not add new things/functionality. The current code on `main` doesn't look at this property, so why do we need to check this property as part of this PR?

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java (hunk 10 above)
Comment (nastra) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:282]: what I mean is that we shouldn't be doing whather it's doing in `throwErrorForExistedToContent()` at this point and just throw the `AlreadyExistsException` as it was done previously. We don't want do add new functionality in this PR IMO

**Comment by nk1506** [comment]
@nastra , Please let me know if there are more comments ?

File: core/src/main/java/org/apache/iceberg/BaseMetastoreOperations.java [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+   *
+   * @param contentName full name of the content
+   * @param newMetadataLocation the path of the new commit file
+   * @param properties properties for retry
+   * @param newMetadataCheckSupplier check if the latest metadata presents or not using metadata
+   *     location for table, version id for view.
+   * @return Commit Status of Success, Failure or Unknown
+   */
+  protected CommitStatus checkCommitStatus(
+      String contentName,
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/BaseMetastoreOperations.java:62]: it's not clear what `contentName` refers to. So I would rename this to `tableOrViewName` (and also reflect this in the javadoc in L54

File: core/src/main/java/org/apache/iceberg/BaseMetastoreOperations.java [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+      String newMetadataLocation,
+      Map<String, String> properties,
+      Supplier<Boolean> newMetadataCheckSupplier) {
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/BaseMetastoreOperations.java:65]:
```suggestion
      Supplier<Boolean> commitStatusSupplier) {
```

File: core/src/main/java/org/apache/iceberg/BaseMetastoreTableOperations.java [hunk 14]
```diff
@@ -291,6 +284,8 @@ public long newSnapshotId() {
     };
   }
 
+  /** @deprecated Use {@link BaseMetastoreOperations.CommitStatus} */
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/BaseMetastoreTableOperations.java:287]:
```suggestion
  /** @deprecated since 1.6.0, will be removed in 1.7.0; use {@link BaseMetastoreOperations.CommitStatus} instead */
```

File: core/src/main/java/org/apache/iceberg/BaseMetastoreTableOperations.java [hunk 15]
```diff
@@ -309,65 +304,39 @@ protected enum CommitStatus {
...
+            .name());
+  }
+
+  /**
+   * Checks the new metadata location presents or not after refreshing the table.
+   *
+   * @param newMetadataLocation newly written metadata location
+   * @return true if the new metadata location presents with current or previous metadata files.
+   */
+  protected boolean checkCurrentMetadataLocation(String newMetadataLocation) {
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/BaseMetastoreTableOperations.java:322]: why does this have to be protected?

File: core/src/main/java/org/apache/iceberg/BaseMetastoreTableOperations.java (hunk 8 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/BaseMetastoreTableOperations.java:324]: this hasn't been addressed yet

File: core/src/main/java/org/apache/iceberg/BaseMetastoreTableOperations.java [hunk 16]
```diff
@@ -309,65 +304,39 @@ protected enum CommitStatus {
...
+                config.properties(),
+                () -> checkCurrentMetadataLocation(newMetadataLocation))
+            .name());
+  }
+
+  /**
+   * Checks the new metadata location presents or not after refreshing the table.
+   *
+   * @param newMetadataLocation newly written metadata location
+   * @return true if the new metadata location presents with current or previous metadata files.
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/BaseMetastoreTableOperations.java:320]:
```suggestion
   * @return true if the new metadata location is the current metadata location or present within previous metadata files.
```

please also update the javadoc in L317 with a similar sentence as here

File: core/src/main/java/org/apache/iceberg/CatalogUtil.java (hunk 2 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/CatalogUtil.java:149]: this hasn't been addressed yet

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 17]
```diff
@@ -273,6 +282,21 @@ public void renameTable(TableIdentifier from, TableIdentifier originalTo) {
     }
   }
 
+  private void validateTable(
```
Comment (nastra) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:285]: this feels like it should be named `validateTableIsIcebergTableOrView`

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 18]
```diff
@@ -222,9 +211,29 @@ public boolean dropTable(TableIdentifier identifier, boolean purge) {
...
+        .filter(
+            table ->
+                table.getParameters() != null
+                    && tableTypeProp.equalsIgnoreCase(
+                        table.getParameters().get(BaseMetastoreTableOperations.TABLE_TYPE_PROP)))
+        .map(table -> TableIdentifier.of(namespace, table.getTableName()))
+        .collect(Collectors.toList());
+  }
+
+  private void renameContent(
```
Comment (nastra) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:232]:
```suggestion
  private void renameTableOrView(
```

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 19]
```diff
@@ -222,9 +211,29 @@ public boolean dropTable(TableIdentifier identifier, boolean purge) {
...
+      TableIdentifier from,
+      TableIdentifier originalTo,
+      HiveOperationsBase.ContentType contentType) {
+    Preconditions.checkArgument(isValidIdentifier(from), "Invalid identifier: %s", from);
```
Comment (nastra) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:236]: this is changing existing behavior. Previously it would throw `NoSuchTableException`

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 20]
```diff
@@ -250,7 +259,7 @@ public void renameTable(TableIdentifier from, TableIdentifier originalTo) {
             return null;
           });
 
-      LOG.info("Renamed table from {}, to {}", from, to);
+      LOG.info("Renamed {} from {}, to {}", contentType.name(), from, to);
```
Comment (nastra) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:262]: should this not be using `contentType.value()` instead of `name()`?

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveOperationsBase.java [hunk 21]
```diff
@@ -139,22 +212,40 @@ static StorageDescriptor storageDescriptor(TableMetadata metadata, boolean hiveE
       storageDescriptor.setInputFormat("org.apache.hadoop.mapred.FileInputFormat");
       serDeInfo.setSerializationLib("org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe");
     }
+
     storageDescriptor.setSerdeInfo(serDeInfo);
     return storageDescriptor;
   }
 
-  static void cleanupMetadata(FileIO io, String commitStatus, String metadataLocation) {
+  default void cleanupMetadataAndUnlock(
```
Comment (nastra) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveOperationsBase.java:220]: I would probably just leave `cleanupMetadata` as it is and introduce `cleanupMetadataAndUnlock()` as a separate method

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveOperationsBase.java [hunk 22]
```diff
@@ -139,22 +212,40 @@ static StorageDescriptor storageDescriptor(TableMetadata metadata, boolean hiveE
...
+  default HiveLock lockObject(
+      Map<String, String> properties, Configuration conf, String catalogName) {
+    if (hiveLockEnabled(conf, properties)) {
+      return new MetastoreLock(conf, metaClients(), catalogName, database(), table());
+    } else {
+      return new NoLock();
+    }
+  }
+
+  default Table newHmsTable(Map<String, String> properties) {
```
Comment (nastra) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveOperationsBase.java:246]: why do we need to change this method's parameter?

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveOperationsBase.java (hunk 6 above)
Comment (nastra) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveOperationsBase.java:277]: can we revert any changes in this class starting from L276 to L473? I feel like we're extracting too much functionality into a common place and some things are slightly different for tables/views later.
That's why I think it would be better to revert L276 to L473 at this point. In the first iteration of view support it's probably ok if some things are duplicated and once View support for Hive is in, we can do another round of refactoring to extract what's really common

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 7 above)
Comment (nastra) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:174]: see my comment in https://github.com/apache/iceberg/pull/10001/files#r1533430710. Please revert these changes here. Once view support is in, we can revisit what we'll extract out

File: core/src/main/java/org/apache/iceberg/CatalogUtil.java (hunk 2 above)
Comment (nk1506) [reply core/src/main/java/org/apache/iceberg/CatalogUtil.java:149]: My Bad, I missed to push the corresponding commit.

File: core/src/main/java/org/apache/iceberg/BaseMetastoreTableOperations.java (hunk 15 above)
Comment (nk1506) [reply core/src/main/java/org/apache/iceberg/BaseMetastoreTableOperations.java:322]: Not `Private` , bcoz It should be available outside of the core-package. Not `public` as it should be only scoped too child class. Do we need `public` privilege here ?

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java (hunk 19 above)
Comment (nk1506) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:236]: Intent here was to have common exception for `Table/View`. But I think with `View` impl we will have to revisit this.

File: core/src/main/java/org/apache/iceberg/BaseMetastoreTableOperations.java (hunk 15 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/BaseMetastoreTableOperations.java:322]: I think this should be `private`. Why would it need to be available for subclasses? I don't think subclasses would call this code?

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 23]
```diff
@@ -240,9 +239,9 @@ protected void doCommit(TableMetadata base, TableMetadata metadata) {
             tbl, updateHiveTable, hiveLockEnabled(metadata, conf) ? null : baseMetadataLocation);
         lock.ensureActive();
 
-        commitStatus = CommitStatus.SUCCESS;
+        commitStatus = BaseMetastoreOperations.CommitStatus.SUCCESS;
```
Comment (nk1506) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:242]: Since we are depreciating `BaseMetastoreTableOperations.CommitStatus` .

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 24]
```diff
@@ -304,30 +305,16 @@ protected void doCommit(TableMetadata base, TableMetadata metadata) {
...
-  @VisibleForTesting
-  Table loadHmsTable() throws TException, InterruptedException {
-    try {
-      return metaClients.run(client -> client.getTable(database, tableName));
-    } catch (NoSuchObjectException nte) {
-      LOG.trace("Table not found {}", fullName, nte);
-      return null;
-    }
-  }
+  public void setHmsParameters(
```
Comment (nk1506) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:315]: It has been split into two parts , one is specific to Table like `snapshot,stats, etc.` Other pat is common parameters as ` schema, location, uuid. `

File: core/src/main/java/org/apache/iceberg/BaseMetastoreTableOperations.java (hunk 15 above)
Comment (nk1506) [reply core/src/main/java/org/apache/iceberg/BaseMetastoreTableOperations.java:325]: It depends how the [checkCommitStatus](https://github.com/apache/iceberg/pull/10001/files#diff-e502621d52f86cf0ec3187dda30ac61f6b76efb7b6276bc8d233ccb2c836fb98R284) is being called. It it is using 
`BaseMetastoreTableOperations.checkCommitStatus`  it can be private . it it is using `BaseMetastoreOperations.checkCommitStatus` it should be protected.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 25]
```diff
@@ -282,7 +281,12 @@ protected void doCommit(TableMetadata base, TableMetadata metadata) {
             database,
             tableName,
             e);
-        commitStatus = checkCommitStatus(newMetadataLocation, metadata);
+        commitStatus =
```
Comment (nk1506) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:284]: Here we can use the original method from `BaseMetastoreTableOperations.checkCommitStatus`  too.

File: core/src/main/java/org/apache/iceberg/BaseMetastoreOperations.java [hunk 26]
```diff
@@ -0,0 +1,118 @@
...
+
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+import org.apache.iceberg.util.PropertyUtil;
+import org.apache.iceberg.util.Tasks;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BaseMetastoreOperations {
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/BaseMetastoreOperations.java:38]:
```suggestion
public abstract class BaseMetastoreOperations {
```

File: core/src/test/java/org/apache/iceberg/TestCatalogUtil.java [hunk 27]
```diff
@@ -223,6 +223,32 @@ public void loadCustomMetricsReporter_badClass() {
         .hasMessageContaining("does not implement MetricsReporter");
   }
 
+  @Test
+  public void fullTableNameWithDifferentValues() {
+    String uriTypeCatalogName = "thrift://host:port/db.table";
+    String nameSpaceWithOneLevel = "ns";
+    String nameSpaceWithTwoLevel = "ns.l2";
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestCatalogUtil.java:230]:
```suggestion
    String namespaceWithTwoLevels = "ns.l2";
```

File: core/src/test/java/org/apache/iceberg/TestCatalogUtil.java [hunk 28]
```diff
@@ -223,6 +223,32 @@ public void loadCustomMetricsReporter_badClass() {
         .hasMessageContaining("does not implement MetricsReporter");
   }
 
+  @Test
+  public void fullTableNameWithDifferentValues() {
+    String uriTypeCatalogName = "thrift://host:port/db.table";
+    String nameSpaceWithOneLevel = "ns";
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestCatalogUtil.java:229]:
```suggestion
    String namespace = "ns";
```

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 29]
```diff
@@ -239,7 +250,7 @@ public void renameTable(TableIdentifier from, TableIdentifier originalTo) {
 
     try {
       Table table = clients.run(client -> client.getTable(fromDatabase, fromName));
-      HiveOperationsBase.validateTableIsIceberg(table, fullTableName(name, from));
+      validateTableIsIcebergTableOrView(contentType, name, table, from);
```
Comment (nastra) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:253]: rather than passing the catalog name / table / from separately, it's probably better to just pass the table and the fullTableName: `fullTableName(name, from)`

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 30]
```diff
@@ -273,6 +284,21 @@ public void renameTable(TableIdentifier from, TableIdentifier originalTo) {
     }
   }
 
+  private void validateTableIsIcebergTableOrView(
+      HiveOperationsBase.ContentType contentType,
+      String catalogName,
```
Comment (nastra) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:289]:
```suggestion
```

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveOperationsBase.java [hunk 31]
```diff
@@ -62,6 +82,48 @@ interface HiveOperationsBase {
 
   String table();
 
+  String catalogName();
```
Comment (nastra) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveOperationsBase.java:85]: I don't think we should be introducing `catalogName()`/ `contentType()` to this API just for printing it in a `TRACE` log inside `loadHmsTable()`. Please remove `catalogName()`/ `contentType()` from this API

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveOperationsBase.java [hunk 32]
```diff
@@ -62,6 +82,48 @@ interface HiveOperationsBase {
...
+
+  ContentType contentType();
+
+  default Table loadHmsTable() throws TException, InterruptedException {
+    try {
+      return metaClients().run(client -> client.getTable(database(), table()));
+    } catch (NoSuchObjectException nte) {
+      LOG.trace(
+          "{} not found {}", contentType(), catalogName() + "." + database() + "." + table(), nte);
```
Comment (nastra) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveOperationsBase.java:94]:
```suggestion
          "{} not found {}", contentType(), database() + "." + table(), nte);
```

this is just TRACE logging, so it should be ok to not have the catalog. I'm not even sure that logging stmt should even exist

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveOperationsBase.java [hunk 33]
```diff
@@ -62,6 +82,48 @@ interface HiveOperationsBase {
...
+      String newMetadataLocation,
+      Schema schema,
+      String uuid,
+      Set<String> obsoleteProps,
+      Supplier<String> previousLocationSupplier) {
+    Map<String, String> parameters =
+        Optional.ofNullable(tbl.getParameters()).orElseGet(Maps::newHashMap);
+
+    if (!obsoleteProps.contains(TableProperties.UUID) && uuid != null) {
+      parameters.put(TableProperties.UUID, uuid);
```
Comment (nastra) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveOperationsBase.java:111]: In https://github.com/apache/iceberg/pull/9852/files#diff-db46657b84d66e084e15f31b8dab21577efb2ae7102863f94c6c9477782de676R206 you're passing the View UUID and I wonder whether this is actually correct to do so?

This is another example where it can be dangerous to have a common method to set something that seems could be common across tables/views.

I would suggest to remove `setCommonHmsParameters()` from this API

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveOperationsBase.java [hunk 34]
```diff
@@ -76,11 +138,11 @@ default boolean exposeInHmsProperties() {
     return maxHiveTablePropertySize() > 0;
   }
 
-  default void setSchema(TableMetadata metadata, Map<String, String> parameters) {
```
Comment (nastra) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveOperationsBase.java:79]: we need to keep the original method (for backwards compatibiliy), but the method can refer to the new `setSchema()` method:

```
default void setSchema(TableMetadata metadata, Map<String, String> parameters) {
    setSchema(metadata.schema(), parameters);
  }
```

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveOperationsBase.java [hunk 35]
```diff
@@ -123,13 +185,14 @@ default void persistTable(Table hmsTable, boolean updateHiveTable, String metada
     }
   }
 
-  static StorageDescriptor storageDescriptor(TableMetadata metadata, boolean hiveEngineEnabled) {
```
Comment (nastra) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveOperationsBase.java:126]: same as above, we need to keep `StorageDescriptor storageDescriptor(TableMetadata metadata, boolean hiveEngineEnabled)` but it can internally call `static StorageDescriptor storageDescriptor(Schema schema, String location, boolean hiveEngineEnabled)`

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 36]
```diff
@@ -304,30 +305,16 @@ protected void doCommit(TableMetadata base, TableMetadata metadata) {
...
-      return metaClients.run(client -> client.getTable(database, tableName));
-    } catch (NoSuchObjectException nte) {
-      LOG.trace("Table not found {}", fullName, nte);
-      return null;
-    }
-  }
+  public void setHmsParameters(
+      TableMetadata metadata, Table tbl, String newMetadataLocation, Set<String> obsoleteProps) {
 
-  private void setHmsTableParameters(
```
Comment (nastra) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:324]: please revert these changes

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 24 above)
Comment (nastra) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:315]: please revert this to how it was before

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 37]
```diff
@@ -365,6 +353,10 @@ private void setHmsTableParameters(
     }
 
     // Set the basic statistics
+    Map<String, String> summary =
```
Comment (nastra) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:356]: why is this change needed?

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 38]
```diff
@@ -457,6 +448,16 @@ public String table() {
     return tableName;
   }
 
+  @Override
+  public String catalogName() {
```
Comment (nastra) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:452]: please remove these two methods

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCatalog.java [hunk 39]
```diff
@@ -1041,7 +1041,7 @@ public void testNotExposeTableProperties() {
         .doesNotContainKey(CURRENT_SNAPSHOT_ID)
         .doesNotContainKey(CURRENT_SNAPSHOT_TIMESTAMP);
 
-    ops.setSchema(metadata, parameters);
+    ops.setSchema(metadata.schema(), parameters);
```
Comment (nastra) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCatalog.java:1044]:
```suggestion
    ops.setSchema(metadata, parameters);
```

Comment (nastra) [reply hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCatalog.java:1044]: no need to change this, as we need to keep the original method signature

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 40]
```diff
@@ -304,23 +310,13 @@ protected void doCommit(TableMetadata base, TableMetadata metadata) {
       throw new CommitFailedException(e);
 
     } finally {
-      cleanupMetadataAndUnlock(commitStatus, newMetadataLocation, lock);
+      HiveOperationsBase.cleanupMetadataAndUnlock(io(), commitStatus, newMetadataLocation, lock);
```
Comment (nastra) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:313]:
```suggestion
      cleanupMetadataAndUnlock(io(), commitStatus, newMetadataLocation, lock);
```
I don't think you need to qualify this method

Comment (nk1506) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:313]: It expects interface with static method . Plz let me know if you think other way around. 
Screenshot 2024-03-25 at 10 21 00 PM

**Review by nastra — APPROVED** [review]
